Add module for dispersion image calculation

ariellellouch opened the pull request:
This module was created to generate dispersion images directly from patches. It uses the distance coordinates to compute a dispersion image using the phase-shift method. 

## Checklist

I have (if applicable):

- [X] referenced the GitHub issue this PR closes.
- [X] documented the new feature with docstrings or appropriate doc page.
- [X] included a test. See [testing guidelines](https://dascore.org/contributing/testing.html).
- [X] your name has been added to the contributors page (docs/contributors.md).
- [X] added the "ready_for_review" tag once the PR is ready to be reviewed.

d-chambers commented:
Hey @ariellellouch, great start! The docstring looks comprehensive and the method and its citations look solid. I will try to find some time in the next few days for a proper review (which means brushing up on the method), but after a quick glance there are a couple of things you can work on in the meantime:

1. testing

We need to ensure DASCore's test suite exercises the new code, preferably every line. To do that you would create a new python file (`tests/test_transform/test_dispersion.py`). The simplest test file in the transforms module you can reference as an example is `test_fft.py`. Then you can run your tests with pytest like: `pytest tests/test_transforms/test_dispersion.py`.  Let me know if you need any help with this, I know pytest (and automated testing in general) can be a bit of a learning curve if you are new to it.

2.  `DispersionParameterError` is probably a little too specific. I would recommend just using `ParameterError` which is already in the exceptions module.

3. If your transform requires dimensions "time" and "distance" the easiest way to ensure this is the case is to use the `required_dims` parameter of [`patch_function`](https://dascore.org/api/dascore/utils/patch/patch_function.html)

e.g., 

```python 
@patch_function(required_dims=("time", "distance"))
def phase_shift(
    patch: PatchType, vels: Sequence[float], direction: str, approxdf: float = 0.0
) -> PatchType:
```

Great work so far.

ariellellouch commented:
Hi @d-chambers,

I added a basic testing unit, as well as fixed your other comments. Sorry if this is amateur coding, but I did my best 

Ariel

d-chambers reviewed (changes requested):
Hey @ariellellouch,

I finally got a chance to go through this, sorry for the delay. There are a few things in the comments I have listed below.  I tried to list out not only the suggested changes but the reasoning for them. Let me know if you have any questions and sorry if I am asking some dumb questions, I am new to MASW.

I am also a bit concerned about testing the correctness of the output, not just the shape and coordinates, but we don't have a dispersive example patch. We should look into making a synthetic one (probably in a different PR), then ensuring the peaks are in the right places. 

If you have too much going on right now (which is completely understandable) I can take a few minutes towards the end of the week and push this over the finish line. Whatever works for you. 

Thanks to @ahmadtourei who helped me do the review.

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+        f = np.arange(approxnf) * fs / (approxnf - 1)
+    else:
+        f = np.arange(nt) * fs / (nt - 1)
+    f[1] - f[0]
```
Done

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+
+@patch_function(required_dims=("time", "distance"))
+def phase_shift(
+    patch: PatchType, vels: Sequence[float], direction: str, approxdf: float = -1.0
```
Done. Assuming I understood the third point (I set approx_resolution to None by default)

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+    patch: PatchType, vels: Sequence[float], direction: str, approxdf: float = -1.0
+) -> PatchType:
+    """
+    Compute dispersion images using the phase-shift method (Park et al., 1999)
+    Inspired by https://geophydog.cool/post/masw_phase_shift/.
```
Done

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+    dt = (time[1] - time[0]) / np.timedelta64(1, "s")  # There has to be an easier way.
+    nchan = dist.size
+    nt = time.size
+    m, n = patch_cop.data.shape
+    assert m == nchan
+    assert n == nt
```
Done

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+    - Dims/Units of the output are forced to be 'frequency' ('Hz')
+      and 'velocity' ('m/s')
+
+    - The patch's distance coordinates are assumed to be ordered by
+    distance from the source, and not "fiber distance". In other
+    words, data are effectively mapped along a 2-D line.
```
No. The phase method works in a "relative" way - if we start at distance X, the phase shift of the first channel will be non-zero, but it's not a problem.

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+    vels
+        NumPY array of positive velocities, monotonically increasing, for
+        which the dispersion will be computed
+    direction
+        Indicating whether the event is down-dip ('ltr') or up-dip ('rtl')
```
You are right, actually. We should have the user order the data the right way. 
Deleted this option.

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+    for ci in range(nv):
+        for fi in range(hnf):
+            fc[ci, fi] = abs(
+                sum(np.exp(preamb * w[fi] / vels[ci] * dist) * fft_d[:, fi])
```
Good catch! 
done

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+
+
+@patch_function(required_dims=("time", "distance"))
+def phase_shift(
```
Done

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+                sum(np.exp(preamb * w[fi] / vels[ci] * dist) * fft_d[:, fi])
+            )
+
+    attrs = dict(category="Dispersion")
```
Done

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+
+    if direction == "ltr":
+        dirflag = 1
+    elif direction == "rtl":
+        dirflag = -1
+    else:
+        raise ParameterError("Direction can only be ltr or rtl")
```
Deleted this flag (previous comment)

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+    disp_patch.viz.waterfall(show=True, scale=0.5, ax=ax)
+
+    """
+    patch_cop = patch.new()
```
Thanks, done

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+    disp_patch = patch.new(
+        data=fc, coords=coords, attrs=attrs, dims=["velocity", "frequency"]
+    )
+    disp_patch.set_units(velocity="m/s", frequency="Hz")
+
+    return disp_patch
```
Done

ariellellouch commented on tests/test_transform/test_dispersion.py:
```diff
@@ -0,0 +1,38 @@
+
+    def test_dispersion(self, dispersion_patch):
+        """Check consistency of test_dispersion module."""
+        dims = dispersion_patch.dims
+        dimvel = [x.startswith("vel") for x in dims]
+        dimfreq = [x.startswith("freq") for x in dims]
+
+        assert any(dimvel)
+        assert any(dimfreq)
```
"frequency" and not "distance", but done :)

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+    else:
+        raise ParameterError("Direction can only be ltr or rtl")
+
+    if patch.dims[0] == "time":
+        patch_cop.transpose("distance", "time")
```
Done.

ariellellouch commented on dascore/transform/dispersion.py:
```diff
@@ -0,0 +1,137 @@
+    hnf = int(nf / 2)
+    fc = np.zeros(shape=(nv, hnf))
+    preamb = 1j * dirflag
+    for ci in range(nv):
+        for fi in range(hnf):
+            fc[ci, fi] = abs(
+                sum(np.exp(preamb * w[fi] / vels[ci] * dist) * fft_d[:, fi])
```
I am leaving this open and will try something myself.
One thing that we could consider is numba, but I don't know if you like the dependency.

d-chambers commented:
Good work, it looks much better.

> I have a synthetic dataset ready for that, but the question is whether I should open a new PR or piggyback on this one? It's not trivial how to compare to exact results because there are numerical resolution issues.

If you are up to it, adding it here is probably better. That way the example can actually use a patch which has some dispersion and make a nicer plot.  I think the steps for you will more or less be these:

1. Create a patch from your synthetic data
2. Save the patch to disk using dasdae format (we will add other writing formats later on)
3. clone the [dasdae test_data repo](https://github.com/DASDAE/test_data)
4. add your data files there, and commit it. Then push either to the master branch, or make another branch and open a PR if you can't
5. calculate file hash using [`pooch.file_hash`](https://www.fatiando.org/pooch/latest/api/generated/pooch.file_hash.html)
6. Create a new entry in dascore's data registry (dascore/data_registry.txt)
7. Create a new patch example function in dascore/examples.py. You can use [example_event_1](https://github.com/DASDAE/dascore/blob/12e2abf3fec9f18bc17e9bdbfc5bbdea64f9f524/dascore/examples.py#L172-L180) as an example. 
8. Update your docstring in `dispersion_phase_shift` to load and use this new patch. For example, if you select the name "dispersive_event" just swap that out for "example_event_1". 

[This](https://dascore.org/contributing/adding_test_data.html#adding-a-data-file)  section of the docs also shows how to add an example file.

d-chambers commented:
Also when you generate your example patch its much nicer if you remember to set the units before saving it. I forgot to do that on the "example_event_1" patch which is why you had use the `set_units("mm/(m*s)", distance='m', time='s')` line in your docstring.

d-chambers commented:
I found a few other small things that i just fixed. Mainly, adding tests for all the raised exceptions and deleting the docstring line which described the removed parameter. 

I am probably a little extreme with how much effort I spend trying to get 100% code coverage, but I did find a subtle bug by doing it: 

```python
    if not np.all(np.diff(phase_velocities)) > 0:
```
should be:

```python
    if not np.all(np.diff(phase_velocities) > 0):
```

In the first case the `Parameter` is never raised unless a diff in the array is 0 (some diffs can happily be less than 0).

ariellellouch commented:
This is actually a pretty good catch - sorry I missed the parenthesis.

1) Added the example as you suggested! It works for me, besides a small problem that I don't know how to fix.
This is how I create the patch:

attrs = dict(
    data_type="velocity",
    id="dispersion_data",
    data_units="m/s",
    data_category="DAS"
)

But for some reason the units are not kept and "revert" to mm/ m / s. Maybe you can help elucidate the mystery...

2) I added what I think is important functionality (especially for run-tume) to limit the frequency range in which the dispersion is calculated. Especially for surface waves, you may want let's say 5-30 Hz, but data sampled at 2000 Hz would be pretty painful. So, I added min/max frequency. I tried to follow your tests convention, hope it is OK - feel free to change. 

3) Added more testing to the functionality in (2)

See you at the sprint.
Ariel

d-chambers commented:
> But for some reason the units are not kept and "revert" to mm/ m / s. Maybe you can help elucidate the mystery...

When I do this:
```python
import dascore as dc

dc.get_example_patch("dispersion_event")
```

I get this patch:

![image](https://github.com/DASDAE/dascore/assets/11671536/c0b9f6a8-a215-4040-a31a-a06b9d3d8c8c)

which has data units of "m/s". Is this not the case for you, or are the units getting switched after the dispersion calculations?

ariellellouch commented:
@d-chambers 
Yes, but when I run the example, the units of the transformed dispersion patch are something else , and I have no idea why...

d-chambers commented:
Maybe I am missing something, but this line in the example:

```python       
 .set_units("mm/(m*s)", distance='m', time='s')
````

will set the data units to "mm/(m*s)". Instead, you can just do:

```python       
 .set_units(distance='m', time='s')
````

to leave the data units alone on only set the coordinate units. Does that solve the issue?

ariellellouch commented:
Oh, such a stupid mistake. Sorry about that - I removed all the unnecessary processing steps.

Should be good to review now!

d-chambers commented:
🥳
